plugins/registry.ts

```diff
@@ -826,65 +1029,102 @@ export default class PluginRegistry {
-    // Add a component to the App Bar.
-    // Accepts the following:
-    // - iconUrl - A resolvable URL to use as the button's icon
-    // - action - A function called when the button is clicked, passed the channel and channel member as arguments
-    // - tooltip_text - A string or React element shown for tooltip appear on hover
-    // Returns a unique identifier.
-    registerAppBarComponent(iconUrl, action, tooltipText) {
+    /**
+     * INTERNAL: Subject to change without notice.
+     * Add an item to the App Bar.
+     * @param {string} iconUrl resolvable URL to use as the button's icon.
+     * @param {PluginComponent['action'] | undefined} action called when the button is clicked, passed the channel and channel member as arguments.
+     * @param {React.ReactNode} tooltipText string or React element shown for tooltip appear on hover.
+     * @param {null | string | Array<null | string>} supportedProductIds specifies one or multiple product identifier(s),
+     * identifiers can either be the "real" product uuid, or a product's more commonly accessible plugin id, or '*' to match everything.
+     * @param {PluginComponent | undefined} rhsComponent an optional corresponding RHS component. If provided, its toggler is automatically wired to the action.
+     * @param {ReactResolvable | undefined} rhsTitle the corresponding RHS component's title.
+     * @returns {string} unique identifier
+     */
+    registerAppBarComponent = reArg([
+        'iconUrl',
+        'action',
+        'tooltipText',
+        'supportedProductIds',
+        'rhsComponent',
+        'rhsTitle',
+    ], ({
+        iconUrl,
+        action,
+        tooltipText,
+        supportedProductIds = null,
+        rhsComponent,
+        rhsTitle,
+    }: {
+        iconUrl: string;
+        tooltipText: ReactResolvable;
+        supportedProductIds: ProductScope;
+    } & ({
+        action: PluginComponent['action'];
+        rhsComponent?: never;
+        rhsTitle?: never;
+    } | {
+        action?: never;
+        rhsComponent: PluginComponent;
+        rhsTitle: ReactResolvable;
+    })) => {
         const id = generateId();
 
-        const data = {
-            id,
-            pluginId: this.id,
-            iconUrl,
-            action,
-            tooltipText: resolveReactElement(tooltipText),
-        };
+        const registeredRhsComponent = rhsComponent && this.registerRightHandSidebarComponent({title: rhsTitle, component: rhsComponent});
 
         store.dispatch({
             type: ActionTypes.RECEIVED_PLUGIN_COMPONENT,
             name: 'AppBar',
-            data,
+            data: {
+                id,
+                pluginId: this.id,
+                iconUrl,
+                tooltipText: resolveReactElement(tooltipText),
+                supportedProductIds,
+                ...registeredRhsComponent ? {
+                    action: () => store.dispatch(registeredRhsComponent.toggleRHSPlugin),
```

Comment:
Indeed this is not ideal, but `store` should be reference-safe in the same way that `stores/redux_store` is. (It's also the same pattern that Playbooks [currently uses](https://github.com/mattermost/mattermost-plugin-playbooks/blob/master/webapp/src/index.tsx#L196-L199).) 

This is necessary because `action` somewhat deceivingly named—it's just a callback that receives a channel and channelMember (https://github.com/mattermost/mattermost-webapp/blob/global-apps-bar/components/app_bar/app_bar_plugin_component.tsx#L103).